### DISCUSSION TITLE
Attributes not supported ?

### DISCUSSION THREAD

**Discussion by tamighi** [rectorphp/rector#8550] [Q&A] [answered]
Hello ! I love your library ! 
Unfortunately I am trying to write a first custom rule for attributes ...

```
    public function getNodeTypes(): array
    {
        return [\PhpParser\Node\Attribute::class];
    }
    public function refactor(Node $node) {...}
```

Simply returning the node does not work, as the resolver expect specific types:

https://github.com/rectorphp/rector/blob/main/src/Application/ChangedNodeScopeRefresher.php#L86-L96

I have version 1.0.3

Am I missing something ?

**Comment by samsonasik** [comment, marked as answer]
You can use the `Stmt` that contains the Attribute, eg `PhpParser\Node\Stmt\Class_`, then process its `$node->attrGroups` to get the `PhpParser\Node\Attribute`, then process it.

**Reply by samsonasik** [reply]
see `AnnotationToAttributeRector` example for it https://github.com/rectorphp/rector/blob/c0820093adb3b5068f55686b44ead70988969749/rules/Php80/Rector/Class_/AnnotationToAttributeRector.php#L48

**Reply by tamighi** [reply]
Thanks my savior, I'll look at it!

**Comment by tamighi** [comment]
Since this does not work correctly as it is a more complex change:

```
use ApiPlatform\Metadata\ApiFilter;
use ApiPlatform\Metadata\ApiResource;


    ->withConfiguredRule(
        AnnotationToAttributeRector::class, [
            new AnnotationToAttribute(ApiFilter::class),
            new AnnotationToAttribute(ApiResource::class),
        ]
    )
```

I tried to parse the attributes over this change to make them work. I think I will try to directly make an 'annotationToAttributes' parser for ApiPlatform, and if done I can maybe make a PR here. Do you have any model of annotationToAttributes class that I could look at for it to fit your practices ? Or maybe it is already done somewhere ?

**Reply by samsonasik** [reply]
It depends of what you're trying to do, what **before** and **after**, if there is no rule yet for it, you can create your own custom rule, or if you think there is beneficial to add feature to rector core or extensions, eg rector-symfony, you can create a PR

**Reply by tamighi** [reply]
It would be to transform ApiPlatform annotations to attributes. I will look at the rector-symphony rule and the other cores, thanks !
